src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1386,6 +1413,37 @@ public IList<string> ValidValues
         {
             get
             {
+                if (validValuesGeneratorType != null)
+                {
+                    ValidValuesCacheEntry validValuesCacheEntry;
+                    if (validValuesCache.TryGetValue(validValuesGeneratorType, out validValuesCacheEntry))
+                    {
+                        var currentTime = DateTime.Now;
+                        if (DateTime.Compare(validValuesCacheEntry.cacheEntryCreationTime.AddSeconds(CacheExpiration), currentTime) < 0)
+                        {
+                            // Update expired Entry.
+                            validValuesCacheEntry.validValues = validValuesCacheEntry.validValuesGenerator.GetValidValues()?.ToArray();
+                            validValuesCacheEntry.cacheEntryCreationTime = currentTime;
+                            validValuesCache[validValuesGeneratorType] = validValuesCacheEntry;
+                        }
+                    }
+                    else
+                    {
+                        var validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(validValuesGeneratorType);
+                        var validValues = validValuesGenerator.GetValidValues()?.ToArray();
+                        var cacheEntryCreationTime = DateTime.Now;
+                        validValuesCacheEntry = new ValidValuesCacheEntry(validValuesGenerator, validValues, cacheEntryCreationTime);
+                        validValuesCache.TryAdd(validValuesGeneratorType, validValuesCacheEntry);
+                    }
+                    if (validValuesCacheEntry.validValues == null)
+                    {
+                        throw new ValidationMetadataException(
+                            "ValidateSetGeneratedValidValuesListIsEmpty",
+                            null,
+                            Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+                    }
+                    _validValues = validValuesCacheEntry.validValues;
```

Comment:
@lzybkr I still don't correct tests however please review my next step - Is that the right direction?

I removed valid values cache and  CacheExpiration  from ValidateSet attribute, enhance a generator cache (add cleanup - the idea is from [CoreFX](https://github.com/stephentoub/corefx/blob/9820567e1bb1972502bfe6345963eb0ab99ef3f9/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs)), add an abstract class for generators with valid values cache support.